pythonFiles/unittestadapter/utils.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import enum
+import inspect
+import os
+import pathlib
+import unittest
+from typing import List, Tuple, TypedDict, Union
+
+# Types
+
+
+# Inherit from str so it's JSON serializable.
+class TestNodeTypeEnum(str, enum.Enum):
+    class_ = "class"
+    file = "file"
+    folder = "folder"
+    test = "test"
+
+
+class TestData(TypedDict):
+    name: str
+    path: str
+    type_: TestNodeTypeEnum
+
+
+class TestItem(TestData):
+    id_: str
+    lineno: str
+
+
+class TestNode(TestData):
+    children: "List[TestNode | TestItem]"
+
+
+# Helper functions for data retrieval.
+
+
+def get_test_case(suite):
+    """Iterate through a unittest test suite and return all test cases."""
+    for test in suite:
+        if isinstance(test, unittest.TestCase):
+            yield test
+        else:
+            for test_case in get_test_case(test):
+                yield test_case
+
+
+def get_source_line(obj) -> str:
+    """Get the line number of a test case start line."""
+    try:
+        sourcelines, lineno = inspect.getsourcelines(obj)
+    except:
+        try:
+            # tornado-specific, see https://github.com/microsoft/vscode-python/issues/17285.
+            sourcelines, lineno = inspect.getsourcelines(obj.orig_method)
+        except:
+            return "*"
+
+    # Return the line number of the first line of the test case definition.
+    for i, v in enumerate(sourcelines):
+        if v.strip().startswith(("def", "async def")):
+            return str(lineno + i)
+    return "*"
```

Comment:
```suggestion

    return "*"
```
Just to make it clearer this is the fall-through case.